### PR TITLE
[feat] add port for client status

### PR DESCRIPTION
**Pull request by montaguelhz**
```bash
Get Client Status: [OK]

Id            Kind     Host         Container Id  Status      Port  Aux Info
--            ----     ----         ------------  ------      ----  --------
2ae7016a1160  curvebs  client-host  8b484fb2ed8a  Up 2 weeks  9002  {"user":"curve","volume":"/test2"}
15ea0603042d  curvebs  server-host  0af4ecea35b7  Up 2 weeks  9001  {"user":"curve","volume":"/test1"}
```

### CODE REVIEW

**Comment by montaguelhz** [comment]
@Wine93 @caoxianfei1 PTAL~

File: internal/task/task/common/client_status.go [hunk 1]
```diff
@@ -100,18 +174,25 @@ func NewGetClientStatusTask(curveadm *cli.CurveAdm, v interface{}) (*task.Task,
...
 	var status string
+	var port string
 	t.AddStep(&step.ListContainers{
 		ShowAll:     true,
 		Format:      `"{{.Status}}"`,
 		Filter:      fmt.Sprintf("id=%s", containerId),
 		Out:         &status,
 		ExecOptions: curveadm.ExecOptions(),
 	})
+	t.AddStep(&step2Port{
```
Comment (Wine93) [inline internal/task/task/common/client_status.go:185]: Maybe `step2GetPort` is more meaningful.

Comment (montaguelhz) [reply internal/task/task/common/client_status.go]: done

**Review by Wine93 — APPROVED** [review]
LGTM!

**Comment by Cyber-SiKu** [comment]
mv port to Aux Info is better?

**Comment by montaguelhz** [comment]
> show ip:port is better?just look like:
> 
> ```shell
> Get Client Status: [OK]
> 
> Id            Kind     Host         Container Id  Status      Adrress                 Aux Info
> --            ----     ----         ------------  ------                                                       ----  --------
> 2ae7016a1160  curvebs  client-host  8b484fb2ed8a  Up 2 weeks  xxx.xxx.xxx.xxx:9002  {"user":"curve","volume":"/test2"}
> 15ea0603042d  curvebs  server-host  0af4ecea35b7  Up 2 weeks  xxx.xxx.xxx.xxx:9001  {"user":"curve","volume":"/test1"}
> ```
> 
> @Wine93 Or --detail (-vv) to display more information

done

**Comment by caoxianfei1** [comment]
@montaguelhz Have you passed the test

**Comment by caoxianfei1** [comment]
> > @montaguelhz Have you passed the test
> 
> The six ci tests are passed.

I test it but address part is nil.

File: internal/task/task/common/client_status.go [hunk 2]
```diff
@@ -81,11 +93,73 @@ func (s *step2FormatClientStatus) Execute(ctx *context.Context) error {
...
 		AuxInfo:     client.AuxInfo,
 	})
 	return nil
 }
 
+func (s *step2GetAddress) Execute(ctx *context.Context) error {
+	cmd := ctx.Module().DockerCli().TopContainer(s.containerId)
+	out, err := cmd.Execute(s.execOptions)
+	if err != nil {
+		return nil
```
Comment (caoxianfei1) [inline internal/task/task/common/client_status.go:106]: return err?

File: internal/task/task/common/client_status.go [hunk 3]
```diff
@@ -81,11 +93,73 @@ func (s *step2FormatClientStatus) Execute(ctx *context.Context) error {
...
+		res := reg.Split(lines[1], -1)
+		if len(res) > 1 {
+			pid = res[1]
+		}
+	}
+
+	if len(pid) == 0 {
+		return nil
+	}
+
```
Comment (caoxianfei1) [inline internal/task/task/common/client_status.go:121]: Here is a error occur? handle it.

File: internal/task/task/common/client_status.go (hunk 2 above)
Comment (montaguelhz) [reply internal/task/task/common/client_status.go:106]: thx, I haven't tested the situation that top command goes wrong, but why it throw error

File: internal/task/task/common/client_status.go [hunk 4]
```diff
@@ -81,11 +93,73 @@ func (s *step2FormatClientStatus) Execute(ctx *context.Context) error {
...
+	cli.AddOption("--listening")
+	command, err := cli.String()
+	if err != nil {
+		return nil
+	}
+	cmd = ctx.Module().DockerCli().ContainerExec(s.containerId, command)
+	out, err = cmd.Execute(s.execOptions)
+	if err != nil {
+		return nil
+	}
```
Comment (caoxianfei1) [inline internal/task/task/common/client_status.go:135]: ditto.

File: internal/task/task/common/client_status.go [hunk 5]
```diff
@@ -81,11 +93,73 @@ func (s *step2FormatClientStatus) Execute(ctx *context.Context) error {
...
+		}
+	}
+
+	if len(pid) == 0 {
+		return nil
+	}
+
+	// execute "ss" command in container
+	cli := ctx.Module().Shell().SocketStatistics("")
+	cli.AddOption("--no-header")
```
Comment (caoxianfei1) [inline internal/task/task/common/client_status.go:124]: see https://github.com/opencurve/curveadm/pull/295

**Comment by montaguelhz** [comment]
@caoxianfei1 The cause of the problem is that I executed `ss` in the terminal and it showed `0.0.0.0:9001`, but in curveadm it showed `*:9001`, so I added this to the regular expression and error handling has been added too.

**Comment by montaguelhz** [comment]
The result is 
Id            Kind     Host         Container Id  Status       Address  Aux Info                          
--            ----     ----         ------------  ------       -------  --------                          
2ae7016a1160  curvebs  client-host  8b484fb2ed8a  Up 2 months  *:9002   {"user":"curve","volume":"/test2"}
15ea0603042d  curvebs  server-host  0af4ecea35b7  Up 2 months  *:9001   {"user":"curve","volume":"/test1"}

**Comment by montaguelhz** [comment]
@caoxianfei1 PTAL~

File: internal/task/task/common/client_status.go [hunk 6]
```diff
@@ -81,11 +93,76 @@ func (s *step2FormatClientStatus) Execute(ctx *context.Context) error {
...
+
+	regex, err := regexp.Compile(`^.* ((\d+\.\d+\.\d+\.\d+)|\*:\d+).*pid=` + pid + ".*$")
+	if err == nil {
+		mu := regex.FindStringSubmatch(line)
+		if len(mu) > 1 {
+			return mu[1]
+		}
+	}
+	return ""
+}
```
Comment (caoxianfei1) [inline internal/task/task/common/client_status.go]: newer version ss command will answer `RTNETLINK answers: Invalid argument`. See the pr https://github.com/opencurve/curveadm/commit/5adc604ae3cedd156171da4006670b05fbb04ecd

**Review by caoxianfei1 — APPROVED** [review]

**Pull request merged**